Source/Core/Core/HW/DSPHLE/UCodes/Zelda.h

```diff
@@ -78,23 +80,27 @@ class ZeldaAudioRenderer
   }
 
   // Mixes two buffers together while applying a volume to one of them. The
-  // volume ramps up/down in N steps using the provided step delta value.
-  //
-  // Note: On a real GC, the stepping happens in 32 steps instead. But hey,
-  // we can do better here with very low risk. Why not? :)
-  template <size_t N>
-  s32 AddBuffersWithVolumeRamp(std::array<s16, N>* dst, const std::array<s16, N>& src, s32 vol,
-                               s32 step)
+  // volume ramps up/down in 32 steps using the provided delta value.
+  s32 AddBuffersWithVolumeRamp(MixingBuffer& dst, const MixingBuffer& src, s32 vol, s32 delta)
   {
-    if (!vol && !step)
+    if (!vol && !delta)
       return vol;
 
-    for (size_t i = 0; i < N; ++i)
+    s32 step = delta >> 5;
+    for (size_t i = 0; i < 0x40;)
     {
-      (*dst)[i] += ((vol >> 16) * src[i]) >> 16;
+      dst[i] += ((vol >> 16) * src[i]) >> 16;
```

Comment:
Doesn't doing (vol >> 16) pre-multiplication strip out the fractional part of the volume? That's probably a big source of imprecision as well.

Reply:
Ah, no, because it's 1.31 and not 16.16. So it's just making it 1.15 instead.